cmd/bucket-quota.go

```diff
@@ -140,88 +200,59 @@ func startBucketQuotaEnforcement(ctx context.Context, objAPI ObjectLayer) {
 		case <-time.NewTimer(bgQuotaInterval).C:
 			logger.LogIf(ctx, enforceFIFOQuota(ctx, objAPI))
 		}
-
 	}
 }
 
 // enforceFIFOQuota deletes objects in FIFO order until sufficient objects
 // have been deleted so as to bring bucket usage within quota
 func enforceFIFOQuota(ctx context.Context, objectAPI ObjectLayer) error {
-	// Turn off quota enforcement if data usage info is unavailable.
-	if env.Get(envDataUsageCrawlConf, config.EnableOn) == config.EnableOff {
-		return nil
-	}
-
-	buckets, err := objectAPI.ListBuckets(ctx)
-	if err != nil {
-		return err
-	}
-
-	dataUsageInfo, err := loadDataUsageFromBackend(ctx, objectAPI)
-	if err != nil {
-		return err
-	}
-
-	for _, binfo := range buckets {
-		bucket := binfo.Name
-
-		bui, ok := dataUsageInfo.BucketsUsage[bucket]
-		if !ok {
-			// bucket doesn't exist anymore, or we
-			// do not have any information to proceed.
-			continue
-		}
-
+	for bucket, currUsage := range globalBucketQuotaSys.cache.bucketSizes {
 		// Check if the current bucket has quota restrictions, if not skip it
 		cfg, err := globalBucketQuotaSys.Get(bucket)
 		if err != nil {
 			continue
 		}
-
 		if cfg.Type != madmin.FIFOQuota {
 			continue
 		}
-
-		var toFree uint64
-		if bui.Size > cfg.Quota && cfg.Quota > 0 {
-			toFree = bui.Size - cfg.Quota
-		}
-
-		if toFree == 0 {
+		var toFree int64
+		toFree = int64(currUsage) - int64(cfg.Quota)
+		if toFree <= 0 {
 			continue
 		}
-
-		// Allocate new results channel to receive ObjectInfo.
-		objInfoCh := make(chan ObjectInfo)
-
-		// Walk through all objects
-		if err := objectAPI.Walk(ctx, bucket, "", objInfoCh); err != nil {
-			return err
-		}
-
-		// reuse the fileScorer used by disk cache to score entries by
-		// ModTime to find the oldest objects in bucket to delete. In
-		// the context of bucket quota enforcement - number of hits are
-		// irrelevant.
-		scorer, err := newFileScorer(toFree, time.Now().Unix(), 1)
+		// fetch list of oldest objects in bucket eligible for deletion
+		quotaDelFile := pathJoin(bucketMetaPrefix, bucket, fifoDeletionFile)
```

Comment:
I dont think you can create new files in this manner, what is the purpose? @poornas can you re-implement this to use crawler cache to mark files for deletion?

Reply:
fifo quota enforcement needs to build a list of files sorted by age when bucket usage hits a certain threshold. This way, when bucket exceeds quota, the enforcement routine will kick in and use this list to clear oldest entries until usage is within quota.AB asked that this list should be in its own directory under the .minio.sys/buckets/bucketname hierarchy.

Reply:
we should discuss this further, I think adding more and more files like this is going to make backend cryptic. We should re-use what we have perhaps and re-implement FIFO first to use the crawler framework.

Reply:
we should also see if you can re-use the information present in the bloomfilter here, we are like capturing similar information at multiple places it is not going to be easy to debug things in this manner. 

Lets talk to @klauspost and see how we can combine them for similar purposes.

Reply:
Why are we doing this?

It is extremely superflous to have all of this when it should just be part of the crawler.

I thought we agreed that this was just a temporary solution that, while fundamentally broken, was just a stopgap until we could do this properly?

I might as well stop waiting for versioning and push on with more crawler changes.

Reply:
oh yeah, the bloom filters are linked to the crawler cycles, so we have strict cutoff points where we move to a clean BF. Another crawler cannot hook into them without breaking the synchronization.

Reply:
@klauspost, if crawler can give back sorted list of oldest files in a bucket up to certain size that will work for FIFO enforcement. Per discussion with @abperiasamy , the other approach discussed was to use a callback from the crawler - when a FIFO bucket approaches 80% of configured quota registers to get metadata for the bucket when crawler runs and the FIFO routine then builds its sorted list from this metadata and saves a list of files eligible for deletion in .minio.sys/buckets/bucket hierarchy. This is what has been implemented, however if it is cleaner to handle all this within crawler that would be nicer.